coordinator/client/testconnection.go

```diff
@@ -137,13 +144,17 @@ func (conn *TestConnection) CreateDir(p string) error {
 	} else if err := conn.CreateDir(path.Dir(p)); err != nil {
 		return err
 	}
+	conn.lock.Lock()
 	conn.nodes[p] = nil
+	conn.lock.Unlock()
 	conn.updatewatch(p, EventNodeCreated)
 	return nil
 }
 
 // Exists implements Connection.Exists
 func (conn *TestConnection) Exists(p string) (bool, error) {
+	conn.lock.Lock()
+	defer conn.lock.Unlock()
```

Comment:
let's make these read locks and move it to the line just before conn.nodes[p] below